### PR TITLE
merge dev into main

### PR DESCRIPTION
**Pull request by lusxvr**
- added lr scheduling
- added more seeding
- added val set

### CODE REVIEW

File: models/config.py [hunk 1]
```diff
@@ -40,14 +40,15 @@ class VLMConfig:
...
+    lr_mp: float = 2e-3
+    lr_backbones: float = 1e-4
     data_cutoff_idx: int = None
+    val_ratio: float = 0.05
     batch_size: int = 256
     mmstar_batch_size: int = 32
     eval_in_epochs: bool = True
     epochs: int = 5
-    compile: bool = False
+    compile: bool = True
```
Comment (andimarafioti) [inline models/config.py:51]: is this faster?

File: train.py [hunk 2]
```diff
@@ -99,13 +115,27 @@ def test_mmstar(model, tokenizer, test_loader, device):
             total_examples += len(is_correct)
             if is_correct:
                 correct_predictions += sum(is_correct)
-
+    model.train()
     accuracy = correct_predictions / total_examples if total_examples > 0 else 0
-
     return accuracy
 
+def get_lr(it, max_lr, max_steps):
```
Comment (andimarafioti) [inline train.py:122]: Some docs here would help. What is this? there are tons of schedulers

File: train.py [hunk 3]
```diff
@@ -99,13 +115,27 @@ def test_mmstar(model, tokenizer, test_loader, device):
...
+    min_lr = max_lr * 0.1
+    warmup_steps = max_steps * 0.03
+    # 1) linear warmup for warmup_iters steps
+    if it < warmup_steps:
+        return max_lr * (it+1) / warmup_steps
+    # 2) if it > lr_decay_iters, return min learning rate
+    if it > max_steps:
+        return min_lr
```
Comment (andimarafioti) [inline train.py:130]: I thought cosine decay tends towards 0, not towards 1/10 of the max_lr. Are you sure about this?

File: models/config.py (hunk 1 above)
Comment (lusxvr) [reply models/config.py:51]: Yes, substantially

File: train.py (hunk 3 above)
Comment (lusxvr) [reply train.py:130]: Depends on the implementation, I took the details from Karpathy (GPT2)

File: data/collators.py [hunk 4]
```diff
@@ -2,6 +2,9 @@
 
 class VQACollator(object):  # Visual Question Answering Collator
     def __init__(self, tokenizer, max_length):
+        torch.manual_seed(0)
```
[the remark this answers is not in the document]
Comment (lusxvr) [reply data/collators.py]: This is what I am not 100% sure of tbh, I remember that I had the problem once that I had the seed in the top level file but it did not properly get passed down to the classes I was instantiating. But I changed it back, it is way cleaner this way and does not make a difference in my current tests anyway

**Review by andimarafioti — APPROVED** [review]
LFG!
